### PR TITLE
Add random faults

### LINKED ISSUE #347: Add possibility to inject faults at a random time
## User stories
As a User I want to inject faults at a random time in order to test the system even more.

A possible way to do this is to have one RandomFault class, which acts as something like a facade to normal faults.

## (Wiki) References
- ...

## Acceptance Criteria
- [ ] Create a concept for random faults
- [ ] implement the concept
- [ ] write tests for the concept
- [ ] ...

### PR DESCRIPTION
**Pull request by instant-sky**
Fixes #347 

Adds the possibility to create faults that get injected at a random time. For this, refactoring the fault structure is necessary. It uses now a strategy pattern.

## PR checklist

- [ ] Acceptance criteria fulfilled
- [ ] Additional features are tested
- [ ] Docs (code, wiki & diagrams) updated
- [ ] Breaking changes anounced
- [ ] Added classes that inherit from `BaseModel` to `src.constants.tables`
- [ ] Dev-branch has been merged into local branch to resolve conflicts
- [ ] Tests and linter have passed AFTER local merge
- [ ] Another dev reviewed and approved

### CODE REVIEW

File: src/fault_injector/fault_types/fault.py [hunk 1]
```diff
@@ -27,10 +39,23 @@ def __init__(
...
+        self, configuration: FaultConfiguration
+    ) -> FaultStrategy:
+        """Returns a random or regular strategy, depending on the configuration
+
+        :param configuration: the configuration of the Fault
+        :type configuration: FaultConfiguration
+        :return: the FaultStrategy for this fault
+        :rtype: FaultStrategy
+        """
+        return self.STRATEGY_CLASSES[configuration.strategy]
```
Comment (CR1337) [inline src/fault_injector/fault_types/fault.py]:
```suggestion
        return self.STRATEGY_CLASSES[configuration.strategy]()
```

Comment (CR1337) [reply src/fault_injector/fault_types/fault.py]: Your version returns a class and not an object.

File: src/fault_injector/fault_configurations/fault_configuration.py [hunk 2]
```diff
@@ -12,16 +12,22 @@ class FaultConfiguration(BaseModel):
...
 
         @abstractmethod
         def _make(self, data: dict) -> "FaultConfiguration":
             raise NotImplementedError()
 
-    start_tick = IntegerField(null=False)
-    end_tick = IntegerField(null=False)
+    start_tick = IntegerField(null=True)
+    end_tick = IntegerField(null=True)
+    inject_probability = FloatField(null=True)
```
Comment (CR1337) [inline src/fault_injector/fault_configurations/fault_configuration.py:28]: This is the probability that an injection occurs in one tick right?

To achieve a "good" injection rate one could have to pick a really low value here.

I don't mind if you let it that way but you might consider using something like `injection_probability_per_1000_ticks` like I did in `ScheduleConfiguration`.

File: src/fault_injector/fault_configurations/fault_configuration.py [hunk 3]
```diff
@@ -12,16 +12,22 @@ class FaultConfiguration(BaseModel):
...
+    resolve_probability = FloatField(null=True)
```
Comment (CR1337) [inline src/fault_injector/fault_configurations/fault_configuration.py:29]: Here the same. (See one line above)

**Review by CR1337 — CHANGES_REQUESTED** [review]
Looks good so far. Just some small improvement suggestions :)

File: src/fault_injector/fault_configurations/fault_configuration.py (hunk 2 above)
Comment (instant-sky) [reply src/fault_injector/fault_configurations/fault_configuration.py:28]: That is a good point. I think I will leave it like this for now, if it bothers us in the future I will change it.

File: src/fault_injector/fault_types/fault.py (hunk 1 above)
Comment (instant-sky) [reply src/fault_injector/fault_types/fault.py]: Oh, you are right!

File: src/fault_injector/fault_configurations/fault_configuration.py (hunk 2 above)
Comment (CR1337) [reply src/fault_injector/fault_configurations/fault_configuration.py:28]: ok, good plan

**Review by CR1337 — APPROVED** [review]
:+1:

**Pull request merged**